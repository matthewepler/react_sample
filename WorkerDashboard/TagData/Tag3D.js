import * as THREE from 'three'
import OBJLoader from 'three-obj-loader'
OBJLoader(THREE)
import OBJUrl from 'file-loader!images/case.obj'
import TWEEN from 'tween.js'

const sceneWidth = 250
const sceneHeight = 285

let scene = null
let camera = null
let renderer = null
let mesh = null

export function init3D (targetElement) {
  return new Promise((resolve, reject) => {
    try {
      scene = new THREE.Scene()
      scene.background = new THREE.Color(1, 1, 1)

      camera = new THREE.PerspectiveCamera(75, sceneWidth / sceneHeight, 0.1, 10000)
      camera.position.z = 5

      const spotLight = new THREE.SpotLight(0xffffff)
      spotLight.position.set(100, 1000, 1000)
      spotLight.castShadow = true
      spotLight.shadow.mapSize.width = sceneWidth
      spotLight.shadow.mapSize.height = sceneHeight
      spotLight.shadow.camera.near = 500
      spotLight.shadow.camera.far = 4000
      spotLight.shadow.camera.fov = 30
      scene.add(spotLight)

      const loader = new THREE.OBJLoader()
      loader.load(OBJUrl, (obj) => {
        obj.traverse(function (child) {
          if (child instanceof THREE.Mesh) {
            child.geometry.translate(0, -0.2, 0.15)
          }
        })
        mesh = obj
        mesh.castShadow = true
        mesh.rotation.y = Math.PI
        mesh.rotation.z = Math.PI

        mesh.scale.set(2.5, 2.5, 2.5)
        scene.add(mesh)
      })

      // origin reference object for debugging
      // let geometry = new THREE.BoxGeometry(1, 1, 1)
      // const material = new THREE.MeshBasicMaterial({color: 0x00ff00})
      // const cube = new THREE.Mesh(geometry, material)
      // scene.add(cube)

      renderer = new THREE.WebGLRenderer({ antialias: true })
      renderer.setSize(sceneWidth, sceneHeight)

      document.getElementById(targetElement).appendChild(renderer.domElement)
      document.getElementsByTagName('canvas')[0].style.cssFloat = 'right'

      render3D()
      resolve(true)
    } catch (err) {
      reject(false)
    }
  })
}

export function render3D (rotX, rotY, rotZ) {
  requestAnimationFrame(render3D)
  TWEEN.update()

  // auto rotation each frame, for debugging
  // mesh.rotation.x += 0.001
  // mesh.rotation.y += 0.0005
  // mesh.rotation.z += 0.0015

  renderer.render(scene, camera)
}

function animateVector3 (vectorToAnimate, target, options) {
  options = options || {}

  // get targets from options or set to defaults
  const to = new THREE.Vector3(target.x, target.y, target.z) || THREE.Vector3()
  const easing = TWEEN.Easing.Quadratic.InOut
  const duration = options.duration || 1000

  // create the tween
  const tweenVector3 = new TWEEN.Tween(vectorToAnimate)
    .to({ x: to.x, y: to.y, z: to.z }, duration)
    .easing(easing)
    .onUpdate(function (d) {
      if (options.update) {
        // pass in a function via options for feedback on progress
        options.update(d)
      }
    })
    .onComplete(function () {
      // pass in a function via options for verification of completion
      if (options.callback) options.callback()
    })

  // start the tween
  tweenVector3.start()

  // return the tween in case we want to manipulate it later on
  return {
    x: tweenVector3.x,
    y: tweenVector3.y,
    z: tweenVector3.z
  }
}

export function animateTag (toVec) {
  animateVector3(mesh.rotation, toVec)
}

