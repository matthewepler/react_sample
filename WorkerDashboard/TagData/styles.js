import { StyleSheet } from 'aphrodite/no-important'
import { colors, fonts } from 'styles/variables'

export default StyleSheet.create({
  bottomDweller: {
    position: 'absolute',
    top: '90%'
  },
  default: {
    color: 'black'
  },
  dataGroup: {
    marginBottom: '0.5rem'
  },
  error: {
    color: colors.main
  },
  noiseLevelWrapper: {

  },
  offline: {
    color: 'black',
    opacity: '0.2'
  },
  tagDataWrapper: {
    fontFamily: fonts.main,
    marginTop: '1rem'
  }
})
