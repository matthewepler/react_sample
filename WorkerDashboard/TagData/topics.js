const topics = [
  // events
  'fall',
  'struck_by',
  // motions
  'bad_bend',
  'skip',
  'trip',
  'motionless',
  'start_moving',
  'twist',
  'reach',
  // sensors
  'envHi',
  'envLo',
  'orient',
  'accel',
  'gyro',
  'magno',
  'meta'
]

export const activityIndicators = [
  'accel',
  'gyro',
  'orient'
]

export default topics
