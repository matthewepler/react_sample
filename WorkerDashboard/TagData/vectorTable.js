const vectorTable = {
  1: { x: 3, y: 2, z: 1 },
  2: { x: 3, y: 1, z: 2 },
  3: { x: 2, y: 3, z: 1 },
  4: { x: 2, y: 1, z: 3 },
  5: { x: 1, y: 2, z: 3 },
  6: { x: 1, y: 3, z: 2 },
  7: { x: 3, y: 2, z: -1 },
  8: { x: 3, y: 1, z: -2 },
  9: { x: 2, y: 3, z: -1 },
  10: { x: 2, y: 1, z: -3 },
  11: { x: 1, y: 2, z: -3 },
  12: { x: 1, y: 3, z: -2 },
  13: { x: 3, y: -2, z: -1 },
  14: { x: 3, y: -1, z: -2 },
  15: { x: 2, y: -3, z: -1 },
  16: { x: 2, y: -1, z: -3 },
  17: { x: 1, y: -2, z: -3 },
  18: { x: 1, y: -3, z: -2 },
  19: { x: 3, y: -2, z: 1 },
  20: { x: 3, y: -1, z: 2 },
  21: { x: 2, y: -3, z: 1 },
  22: { x: 2, y: -1, z: 3 },
  23: { x: 1, y: -2, z: 3 },
  24: { x: 1, y: -3, z: 2 },
  25: { x: -3, y: -2, z: 1 },
  26: { x: -3, y: -1, z: 2 },
  27: { x: -2, y: -3, z: 1 },
  28: { x: -2, y: -1, z: 3 },
  29: { x: -1, y: -2, z: 3 },
  30: { x: -1, y: -3, z: 2 },
  31: { x: -3, y: 2, z: 1 },
  32: { x: -3, y: 1, z: 2 },
  33: { x: -2, y: 3, z: 1 },
  34: { x: -2, y: 1, z: 3 },
  35: { x: -1, y: 2, z: 3 },
  36: { x: -1, y: 3, z: 2 },
  37: { x: -3, y: 2, z: -1 },
  38: { x: -3, y: 1, z: -2 },
  39: { x: -2, y: 3, z: -1 },
  40: { x: -2, y: 1, z: -3 },
  41: { x: -1, y: 2, z: -3 },
  42: { x: -1, y: 3, z: -2 },
  43: { x: -3, y: 2, z: -1 },
  44: { x: -3, y: 1, z: -2 },
  45: { x: -2, y: 3, z: -1 },
  46: { x: -2, y: 1, z: -3 },
  47: { x: -1, y: 2, z: -3 },
  48: { x: -1, y: 3, z: -2 }
}

export default vectorTable
