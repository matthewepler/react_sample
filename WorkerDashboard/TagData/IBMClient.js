import IBMClient from 'ibmiotf'
import topics from './topics'

// TODO: this will need to be moved to server
// all incoming topics will be mirrored as web socket connections
// OR we leave it hear and protect the broker from anything other than read actions

const defaultConfig = {
  'org': 'ykq7wp',
  // id must be unique for every client connection
  'id': `webAdminDashboard_${Date.now()}`,
  'domain': 'internetofthings.ibmcloud.com',
  'auth-key': 'a-ykq7wp-opokhmx25k',
  'auth-token': 'MZEGIr1zROG&nn0&Wf'
}

export const initIBMClient = (config) => {
  const IBMAppClient = new IBMClient.IotfApplication(config = defaultConfig)
  return IBMAppClient
}

export const initSubscriptions = (client, id) => {
  topics.forEach(topic => {
    // client.subscribeToDeviceEvents(device_type, deviceId, topic, format)
    // '+' === all, for testing only
    client.subscribeToDeviceEvents('hcs_tag', id || '+', topic, 'json')
  })
}
