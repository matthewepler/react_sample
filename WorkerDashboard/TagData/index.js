import React, { Component } from 'react'
import { css } from 'aphrodite/no-important'
import { initIBMClient, initSubscriptions } from './IBMClient'
import { activityIndicators } from './topics'
import mapRange from 'utils/mapRange'
import random from 'lodash/random'

import Card from 'components/Dashboard/Card'
import DataIconRow from 'components/Dashboard/DataIconRow'
import { animateTag, init3D } from './Tag3D'
import styles from './styles'

import vectorTable from './vectorTable'

export default class TagData extends Component {
  constructor (props) {
    super(props)
    this.state = {
      active: false,
      battery: '-',
      altitude: '-',
      humidity: '-',
      lastHeard: 0,
      micLevel: '500',
      objLoaded: false,
      currOrientation: {x: 0, y: 0, z:0},
      serverConnected: false,
      tagConnected: false,
      tagId: null,
      temperature: '-'
    }
    this.IBMAppClient = null

    this.heartbeatInterval = null
    this.heartbeatTimer = 1000
    this.activeThreshold = 10000 // how long before a tag is considered inactive

    this.micLevelWidth = 100

    this.tag3Did = 'tag3D'

    this.handleBrokerDisconnect = this.handleBrokerDisconnect.bind(this)
  }

  componentDidMount () {
    // TODO: get worker's Tag ID# - API call for this ???
    // TODO: how is this re-rendered if a new worker is called? Tie to store
    // so a change there triggers re-render here. Should receive currWorker...?

    this.IBMAppClient = initIBMClient()
    this.IBMAppClient && this.IBMAppClient.connect()
    this.setIBMHandlers(this.IBMAppClient)

    // 3D Tag
    init3D(this.tag3Did).then(result => {
      !this['_calledComponentWillUnmount'] && this.setState({ objLoaded: result })
    })

    // setInterval(this.updateOrientation, 1000)
  }

  setIBMHandlers (IBMAppClient) {
    IBMAppClient.on('connect', () => {
      !this['_calledComponentWillUnmount'] && this.setState({ serverConnected: true })
      initSubscriptions(IBMAppClient, this.state.tagId) // replace with tagID
      this.heartbeatInterval = setInterval(this.checkHeartbeat(), this.heartbeatTimer)
    })
    IBMAppClient.on('deviceEvent', (deviceType, deviceId, eventType, format, payload) => {
      // payload is an array of integers and needs coercing
      // const data = JSON.parse(String(payload))
      // console.log(deviceId, eventType)
      if (!this.state.tagConnected) {
        !this['_calledComponentWillUnmount'] && this.setState({ tagConnected: true })
      }

      !this['_calledComponentWillUnmount'] && this.setState({ lastHeard: Date.now() })
    })
    IBMAppClient.on('disconnect', this.handleBrokerDisconnect)
  }

  handleBrokerDisconnect () {
    !this['_calledComponentWillUnmount'] && this.setState({ serverConnected: false })
    this.disconnect()
  }

  componentWillUnmount () {
    this.IBMAppClient.disconnect()
    this.disconnect()
  }

  disconnect () {
    clearInterval(this.heartbeatInterval)
  }

  checkErrorStatus (stateItem) {
    return css(
      stateItem ? styles.default : styles.error,
    )
  }

  checkConnectionStatus () {
    return css(
      !this.state.tagConnected || !this.state.serverConnected
        ? styles.offline
        : ''
    )
  }

  calcMicLevel () {
    return mapRange(this.state.micLevel, 0, 1023, 0, this.micLevelWidth)
  }

  checkHeartbeat () {
    const diff = Date.now() - new Date(this.state.lastHeard).getTime()
    if (diff > this.activeThreshold) {
      !this['_calledComponentWillUnmount'] && this.setState({ tagConnected: false })
    }
  }

  updateOrientation (num) {
    num = random(1, 48, false) // for testing only
    animateTag(vectorTable[num])
  }

  render () {
    return (
      <Card title='Tag Data' width='500' height='fixedHeight' marginRight >

        {/* 3D tag */}
        <div id={this.tag3Did} onClick={this.updateOrientation.bind(this)} />

        <div className={css(styles.tagDataWrapper)}>
          {/* Server Connection (IBM MQTT Broker) */}
          <DataIconRow
            iconClass={`fa fa-server 
              ${css(styles.statusIcon)} 
              ${this.checkErrorStatus(this.state.serverConnected)}`}
            spanClass={this.checkErrorStatus(this.state.serverConnected)}
            textString={`Server ${this.state.serverConnected ? 'Connected' : 'Disconnected'}`}
            data-test='server-connection-status'
          />

          {/* Tag Connection */}
          <DataIconRow
            iconClass={`fa fa-link 
              ${css(styles.statusIcon)} 
              ${this.checkErrorStatus(this.state.tagConnected)}`}
            spanClass={this.checkErrorStatus(this.state.tagConnected)}
            textString={`Tag ${this.state.tagConnected ? 'Connected' : 'Disconnected'}`}
            data-test='tag-connection-status'
          />

          {/* Active/Inactive */}
          <DataIconRow
            iconClass={`fa fa-arrows 
              ${css(styles.statusIcon)}  
              ${this.checkErrorStatus(this.state.active)}
              ${this.checkConnectionStatus()}`}
            spanClass={`
              ${this.checkErrorStatus(this.state.active)}
              ${this.checkConnectionStatus()}`}
            textString={`${this.state.active ? 'Active' : 'Inactive'}`}
          />

          {/* Altitude */}
          <DataIconRow
            iconClass={`fa fa-arrow-up 
                ${css(styles.statusIcon)} 
                ${this.checkConnectionStatus()}`}
            spanClass={`
              ${css(styles.default)} 
              ${this.checkConnectionStatus()}`}
            textString={`${this.state.altitude} ft`}
          />

          {/* Temperature */}
          <DataIconRow
            iconClass={`fa fa-thermometer-three-quarters 
                ${css(styles.statusIcon)}
                ${this.checkConnectionStatus()}`}
            spanClass={`
              ${css(styles.default)}
              ${this.checkConnectionStatus()}`}
            textString={`${this.state.temperature} Â°F`}
          />

          {/* Humidity */}
          <DataIconRow
            iconClass={`fa fa-tint 
                ${css(styles.statusIcon)}
                ${this.checkConnectionStatus()}`}
            spanClass={`
              ${css(styles.default)}
              ${this.checkConnectionStatus()}`}
            textString={`${this.state.humidity}%`}
          />

          {/* Battery Level */}
          <DataIconRow
            iconClass={`fa fa-battery-three-quarters 
                ${css(styles.statusIcon)}
                ${this.checkConnectionStatus()}`}
            spanClass={`
              ${css(styles.default)}
              ${this.checkConnectionStatus()}`}
            textString={`${this.state.battery}%`}
          />

          {/* Mic Level */}
          <DataIconRow
            iconClass={`fa fa-volume-up 
                ${css(styles.statusIcon)}
                ${this.checkConnectionStatus()}`}
            spanClass={`
              ${css(styles.default)}
              ${this.checkConnectionStatus()}`}
            positionClass={css(styles.bottomDweller)}
            textString={`  ${this.state.micLevel}dB`}
            addOnElements={
              <svg width={this.micLevelWidth} height='10' xmlns='http://www.w3.org/2000/svg'>
                <line x1='0' y1='5' x2={this.micLevelWidth} y2='5' strokeWidth='1' stroke='grey' />
                <rect x='0' y='0' width={this.calcMicLevel()} height='10' fill='grey' />
              </svg>
            }
          />

        </div>
      </Card>
    )
  }
}
