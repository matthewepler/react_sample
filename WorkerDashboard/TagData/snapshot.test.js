import React from 'react'
import { mount, render, shallow } from 'enzyme'
// import { renderer } from 'react-test-renderer'
import { fromJS } from 'immutable'
import { StyleSheetTestUtils } from 'aphrodite/no-important'
import TagData from './'

describe('TagData container snapshot', () => {
  beforeEach(() => {
    StyleSheetTestUtils.suppressStyleInjection()
  })
  afterEach(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection()
  })

  it('renders the component', () => {
    const tree = renderComponent({}, 'shallow')
    expect(tree).toMatchSnapshot()
  })

  // @MATEO - this depends on a library completing a method successfully. Should
  // that be mocked? 
  it('creates an IBM client for connecting to MQTT broker', () => {
    const tree = renderComponent({}, 'mount')
    expect(tree.instance().IBMAppClient).not.toBeNull()
  })

  it('should display the server connection status', () => {
    const tree = renderComponent({}, 'shallow')
    tree.setState({ serverConnected: false })
    // these are brittle, sorrry. 
    expect(tree.find(sel('server-connection-status')).props().textString)
      .toEqual('Server Disconnected')
    tree.setState({ serverConnected: true })
    expect(tree.find(sel('server-connection-status')).props().textString)
      .toEqual('Server Connected')
  })

  it('should display the tag connection status', () => {
    const tree = renderComponent({}, 'shallow')
    tree.setState({ tagConnected: false })
    // these are brittle, sorrry. 
    expect(tree.find(sel('tag-connection-status')).props().textString)
      .toEqual('Tag Disconnected')
    tree.setState({ tagConnected: true })
    expect(tree.find(sel('tag-connection-status')).props().textString)
      .toEqual('Tag Connected')
  })

  it.skip('should render the data from the tag to the component', () => {
    // to be completed
  })
})

function renderComponent (props = {}, type) {
  const propsToUse = {
    // no default props
    ...props
  }
  switch (type) {
    case 'shallow':
      return shallow(<TagData {...propsToUse} />)
    case 'mount':
      return mount(<TagData {...propsToUse} />)
    default:
      throw new Error('you did not specify a render method')
  }
}

function sel (id) {
  return `[data-test="${id}"]`
}

// testing for IBM stuff too
// testing for Tag3D
