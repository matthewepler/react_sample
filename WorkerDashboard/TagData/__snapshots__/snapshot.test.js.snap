exports[`TagData container snapshot renders the component 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Card
    height="fixedHeight"
    marginRight={true}
    sizeHoriz="small"
    title="Tag Data"
    type="default"
    width="500">
    <div
        id="tag3D"
        onClick={[Function]} />
    <div
        className="tagDataWrapper_1xqiuw2">
        <DataIconRow
            data-test="server-connection-status"
            iconClass="fa fa-server 
                       
                      error_1k5dggg"
            spanClass="error_1k5dggg"
            textString="Server Disconnected" />
        <DataIconRow
            data-test="tag-connection-status"
            iconClass="fa fa-link 
                       
                      error_1k5dggg"
            spanClass="error_1k5dggg"
            textString="Tag Disconnected" />
        <DataIconRow
            iconClass="fa fa-arrows 
                        
                      error_1k5dggg
                      offline_wz1f40"
            spanClass="
                      error_1k5dggg
                      offline_wz1f40"
            textString="Inactive" />
        <DataIconRow
            iconClass="fa fa-arrow-up 
                         
                        offline_wz1f40"
            spanClass="
                      default_53ysj 
                      offline_wz1f40"
            textString="- ft" />
        <DataIconRow
            iconClass="fa fa-thermometer-three-quarters 
                        
                        offline_wz1f40"
            spanClass="
                      default_53ysj
                      offline_wz1f40"
            textString="- °F" />
        <DataIconRow
            iconClass="fa fa-tint 
                        
                        offline_wz1f40"
            spanClass="
                      default_53ysj
                      offline_wz1f40"
            textString="-%" />
        <DataIconRow
            iconClass="fa fa-battery-three-quarters 
                        
                        offline_wz1f40"
            spanClass="
                      default_53ysj
                      offline_wz1f40"
            textString="-%" />
        <DataIconRow
            addOnElements={
                <svg
                    height="10"
                    width={100}
                    xmlns="http://www.w3.org/2000/svg">
                    <line
                        stroke="grey"
                        strokeWidth="1"
                        x1="0"
                        x2={100}
                        y1="5"
                        y2="5" />
                    <rect
                        fill="grey"
                        height="10"
                        width={48.87585532746823}
                        x="0"
                        y="0" />
                </svg>
            }
            iconClass="fa fa-volume-up 
                        
                        offline_wz1f40"
            positionClass="bottomDweller_10spfwm"
            spanClass="
                      default_53ysj
                      offline_wz1f40"
            textString="  500dB" />
    </div>
</Card>,
  "nodes": Array [
    <Card
      height="fixedHeight"
      marginRight={true}
      sizeHoriz="small"
      title="Tag Data"
      type="default"
      width="500">
      <div
            id="tag3D"
            onClick={[Function]} />
      <div
            className="tagDataWrapper_1xqiuw2">
            <DataIconRow
                  data-test="server-connection-status"
                  iconClass="fa fa-server 
                           
                          error_1k5dggg"
                  spanClass="error_1k5dggg"
                  textString="Server Disconnected" />
            <DataIconRow
                  data-test="tag-connection-status"
                  iconClass="fa fa-link 
                           
                          error_1k5dggg"
                  spanClass="error_1k5dggg"
                  textString="Tag Disconnected" />
            <DataIconRow
                  iconClass="fa fa-arrows 
                            
                          error_1k5dggg
                          offline_wz1f40"
                  spanClass="
                          error_1k5dggg
                          offline_wz1f40"
                  textString="Inactive" />
            <DataIconRow
                  iconClass="fa fa-arrow-up 
                             
                            offline_wz1f40"
                  spanClass="
                          default_53ysj 
                          offline_wz1f40"
                  textString="- ft" />
            <DataIconRow
                  iconClass="fa fa-thermometer-three-quarters 
                            
                            offline_wz1f40"
                  spanClass="
                          default_53ysj
                          offline_wz1f40"
                  textString="- °F" />
            <DataIconRow
                  iconClass="fa fa-tint 
                            
                            offline_wz1f40"
                  spanClass="
                          default_53ysj
                          offline_wz1f40"
                  textString="-%" />
            <DataIconRow
                  iconClass="fa fa-battery-three-quarters 
                            
                            offline_wz1f40"
                  spanClass="
                          default_53ysj
                          offline_wz1f40"
                  textString="-%" />
            <DataIconRow
                  addOnElements={
                        <svg
                              height="10"
                              width={100}
                              xmlns="http://www.w3.org/2000/svg">
                              <line
                                    stroke="grey"
                                    strokeWidth="1"
                                    x1="0"
                                    x2={100}
                                    y1="5"
                                    y2="5" />
                              <rect
                                    fill="grey"
                                    height="10"
                                    width={48.87585532746823}
                                    x="0"
                                    y="0" />
                        </svg>
                  }
                  iconClass="fa fa-volume-up 
                            
                            offline_wz1f40"
                  positionClass="bottomDweller_10spfwm"
                  spanClass="
                          default_53ysj
                          offline_wz1f40"
                  textString="  500dB" />
      </div>
</Card>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance":  {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <TagData />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": TagData {
        "IBMAppClient": null,
        "_reactInternalInstance": [Circular],
        "activeThreshold": 10000,
        "context": Object {},
        "handleBrokerDisconnect": [Function],
        "heartbeatInterval": null,
        "heartbeatTimer": 1000,
        "micLevelWidth": 100,
        "props": Object {},
        "refs": Object {},
        "state": Object {
          "active": false,
          "altitude": "-",
          "battery": "-",
          "currOrientation": Object {
            "x": 0,
            "y": 0,
            "z": 0,
          },
          "humidity": "-",
          "lastHeard": 0,
          "micLevel": "500",
          "objLoaded": false,
          "serverConnected": false,
          "tagConnected": false,
          "tagId": null,
          "temperature": "-",
        },
        "tag3Did": "tag3D",
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Card
          height="fixedHeight"
          marginRight={true}
          sizeHoriz="small"
          title="Tag Data"
          type="default"
          width="500">
          <div
                    id="tag3D"
                    onClick={[Function]} />
          <div
                    className="tagDataWrapper_1xqiuw2">
                    <DataIconRow
                              data-test="server-connection-status"
                              iconClass="fa fa-server 
                                   
                                  error_1k5dggg"
                              spanClass="error_1k5dggg"
                              textString="Server Disconnected" />
                    <DataIconRow
                              data-test="tag-connection-status"
                              iconClass="fa fa-link 
                                   
                                  error_1k5dggg"
                              spanClass="error_1k5dggg"
                              textString="Tag Disconnected" />
                    <DataIconRow
                              iconClass="fa fa-arrows 
                                    
                                  error_1k5dggg
                                  offline_wz1f40"
                              spanClass="
                                  error_1k5dggg
                                  offline_wz1f40"
                              textString="Inactive" />
                    <DataIconRow
                              iconClass="fa fa-arrow-up 
                                     
                                    offline_wz1f40"
                              spanClass="
                                  default_53ysj 
                                  offline_wz1f40"
                              textString="- ft" />
                    <DataIconRow
                              iconClass="fa fa-thermometer-three-quarters 
                                    
                                    offline_wz1f40"
                              spanClass="
                                  default_53ysj
                                  offline_wz1f40"
                              textString="- °F" />
                    <DataIconRow
                              iconClass="fa fa-tint 
                                    
                                    offline_wz1f40"
                              spanClass="
                                  default_53ysj
                                  offline_wz1f40"
                              textString="-%" />
                    <DataIconRow
                              iconClass="fa fa-battery-three-quarters 
                                    
                                    offline_wz1f40"
                              spanClass="
                                  default_53ysj
                                  offline_wz1f40"
                              textString="-%" />
                    <DataIconRow
                              addOnElements={
                                        <svg
                                                  height="10"
                                                  width={100}
                                                  xmlns="http://www.w3.org/2000/svg">
                                                  <line
                                                            stroke="grey"
                                                            strokeWidth="1"
                                                            x1="0"
                                                            x2={100}
                                                            y1="5"
                                                            y2="5" />
                                                  <rect
                                                            fill="grey"
                                                            height="10"
                                                            width={48.87585532746823}
                                                            x="0"
                                                            y="0" />
                                        </svg>
                              }
                              iconClass="fa fa-volume-up 
                                    
                                    offline_wz1f40"
                              positionClass="bottomDweller_10spfwm"
                              spanClass="
                                  default_53ysj
                                  offline_wz1f40"
                              textString="  500dB" />
          </div>
</Card>,
        "_debugID": 2,
        "_renderedOutput": <Card
          height="fixedHeight"
          marginRight={true}
          sizeHoriz="small"
          title="Tag Data"
          type="default"
          width="500">
          <div
                    id="tag3D"
                    onClick={[Function]} />
          <div
                    className="tagDataWrapper_1xqiuw2">
                    <DataIconRow
                              data-test="server-connection-status"
                              iconClass="fa fa-server 
                                   
                                  error_1k5dggg"
                              spanClass="error_1k5dggg"
                              textString="Server Disconnected" />
                    <DataIconRow
                              data-test="tag-connection-status"
                              iconClass="fa fa-link 
                                   
                                  error_1k5dggg"
                              spanClass="error_1k5dggg"
                              textString="Tag Disconnected" />
                    <DataIconRow
                              iconClass="fa fa-arrows 
                                    
                                  error_1k5dggg
                                  offline_wz1f40"
                              spanClass="
                                  error_1k5dggg
                                  offline_wz1f40"
                              textString="Inactive" />
                    <DataIconRow
                              iconClass="fa fa-arrow-up 
                                     
                                    offline_wz1f40"
                              spanClass="
                                  default_53ysj 
                                  offline_wz1f40"
                              textString="- ft" />
                    <DataIconRow
                              iconClass="fa fa-thermometer-three-quarters 
                                    
                                    offline_wz1f40"
                              spanClass="
                                  default_53ysj
                                  offline_wz1f40"
                              textString="- °F" />
                    <DataIconRow
                              iconClass="fa fa-tint 
                                    
                                    offline_wz1f40"
                              spanClass="
                                  default_53ysj
                                  offline_wz1f40"
                              textString="-%" />
                    <DataIconRow
                              iconClass="fa fa-battery-three-quarters 
                                    
                                    offline_wz1f40"
                              spanClass="
                                  default_53ysj
                                  offline_wz1f40"
                              textString="-%" />
                    <DataIconRow
                              addOnElements={
                                        <svg
                                                  height="10"
                                                  width={100}
                                                  xmlns="http://www.w3.org/2000/svg">
                                                  <line
                                                            stroke="grey"
                                                            strokeWidth="1"
                                                            x1="0"
                                                            x2={100}
                                                            y1="5"
                                                            y2="5" />
                                                  <rect
                                                            fill="grey"
                                                            height="10"
                                                            width={48.87585532746823}
                                                            x="0"
                                                            y="0" />
                                        </svg>
                              }
                              iconClass="fa fa-volume-up 
                                    
                                    offline_wz1f40"
                              positionClass="bottomDweller_10spfwm"
                              spanClass="
                                  default_53ysj
                                  offline_wz1f40"
                              textString="  500dB" />
          </div>
</Card>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <TagData />,
}
`;
